File: src/TaylorBot.js/taylorbot/database/repositories/YoutubeCheckerRepository.js
'use strict';

const Log = require('../../tools/Logger.js');

class YoutubeCheckerRepository {
    constructor(db) {
        this._db = db;
    }

    async getAll() {
        try {
            return await this._db.any('SELECT * FROM checkers.youtube_checker;');
        }
        catch (e) {
            Log.error(`Getting Youtube Channels: ${e}`);
            throw e;
        }
    }

    async update(playlistId, guildId, channelId, lastVideoId) {
        try {
            return await this._db.oneOrNone(
                `UPDATE checkers.youtube_checker SET last_video_id = $[last_video_id]
                WHERE playlist_id = $[playlist_id] AND guild_id = $[guild_id] AND channel_id = $[channel_id]
                RETURNING *;`,
                {
                    'last_video_id': lastVideoId,
                    'playlist_id': playlistId,
                    'guild_id': guildId,
                    'channel_id': channelId
                }
            );
        }
        catch (e) {
            Log.error(`Updating Youtube for guild ${guildId}, channel ${channelId}, playlistId ${playlistId}: ${e}`);
            throw e;
        }
    }
}

module.exports = YoutubeCheckerRepository;

File: src/TaylorBot.Net/src/TaylorBot.Net.YoutubeNotifier.Infrastructure/YoutubeCheckerRepository.cs
using Dapper;
using Google.Apis.YouTube.v3.Data;
using Microsoft.Extensions.Options;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using TaylorBot.Net.Core.Infrastructure;
using TaylorBot.Net.Core.Infrastructure.Options;
using TaylorBot.Net.Core.Snowflake;
using TaylorBot.Net.YoutubeNotifier.Domain;
using TaylorBot.Net.YoutubeNotifier.Infrastructure.Models;

namespace TaylorBot.Net.YoutubeNotifier.Infrastructure
{
    public class YoutubeCheckerRepository : PostgresRepository, IYoutubeCheckerRepository
    {
        public YoutubeCheckerRepository(IOptionsMonitor<DatabaseConnectionOptions> optionsMonitor) : base(optionsMonitor)
        {
        }

        public async Task<IEnumerable<YoutubeChecker>> GetYoutubeCheckersAsync()
        {
            using (var connection = Connection)
            {
                connection.Open();
                var checkers = await connection.QueryAsync<YoutubeCheckerDto>("SELECT * FROM checkers.youtube_checker;");
                return checkers.Select(checker => new YoutubeChecker(
                    guildId: new SnowflakeId(checker.guild_id),
                    channelId: new SnowflakeId(checker.channel_id),
                    playlistId: checker.playlist_id,
                    lastVideoId: checker.last_video_id
                ));
            }
        }

        public async Task UpdateLastPostAsync(YoutubeChecker youtubeChecker, PlaylistItemSnippet youtubePost)
        {
            using (var connection = Connection)
            {
                connection.Open();
                await connection.ExecuteAsync(
                    @"UPDATE checkers.youtube_checker SET last_video_id = @LastVideoId
                      WHERE playlist_id = @PlaylistId AND guild_id = @GuildId AND channel_id = @ChannelId;",
                    new
                    {
                        PlaylistId = youtubeChecker.PlaylistId,
                        GuildId = youtubeChecker.GuildId.ToString(),
                        ChannelId = youtubeChecker.ChannelId.ToString(),
                        LastVideoId = youtubePost.ResourceId.VideoId
                    }
                );
            }
        }
    }
}
